cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java

```diff
@@ -413,5 +434,40 @@ private void discardAfterTx(TransactionAware txAware) {
       }
     }
   }
+
+  /**
+   * Cache key the wrap around {@link DatasetCacheKey}. Unlike the {@link DatasetCacheKey}, the comparison includes the
+   * {@link DatasetCacheKey#accessType}.
+   */
+  private static final class AccessAwareDatasetCacheKey {
+
+    private final DatasetCacheKey key;
+
+    private AccessAwareDatasetCacheKey(DatasetCacheKey key) {
+      this.key = key;
+    }
+
+    DatasetCacheKey getDatasetCacheKey() {
+      return key;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      AccessAwareDatasetCacheKey other = (AccessAwareDatasetCacheKey) o;
+      return Objects.equals(key, other.key) && Objects.equals(key.getAccessType(), other.key.getAccessType());
+    }
+
+    @Override
+    public int hashCode() {
+      return 31 * key.hashCode() + key.getAccessType().hashCode();
```

Comment:
Since there is no null, this is more lightweight than `Objects.hash` as it avoid creating an `Object[]` and this method is called every time when hitting the cache, so good to be lightweight.